interfaces/zookeeper_client/v0/schema.py

```diff
@@ -0,0 +1,92 @@
+"""This file defines the schemas for the provider and requirer sides of the zookeeper_client interface.
+
+It must expose two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+"""
+
+from typing import Annotated
+
+from interface_tester.schema_base import DataBagSchema
+from pydantic import (
+    AfterValidator,
+    BaseModel,
+    Field,
+    IPvAnyAddress,
+    conint,
+)
+
+
+class URI(BaseModel):
+    ip: IPvAnyAddress
+    port: conint(ge=0, le=65535) | None
+
+    def __init__(self, value: str) -> None:
+        ip, _, port = value.partition(":")
+        if not port:
+            port = None
+        super().__init__(ip=ip, port=port)
+
+
+def parse_endpoints(value: str):
+    uris, _, znode = value.partition("/")
+    [URI(val) for val in uris.split(",")]
+
+
+Endpoints = Annotated[str, AfterValidator(parse_endpoints)]
```

Comment:
**suggestion:** Perhaps here it might be OK to loosen the typing from `URI` to just be `string`.

Then, we can add a validator in the class itself, something like:

```python
from pydantic import validator
from ipaddress

class ZookeeperProviderAppData(BaseModel):
    endpoints: string = Field(
        description="A comma-seperated list of ZooKeeper server uris, and zNode",
        examples=["10.141.78.133:2181,10.141.78.50:2181,10.141.78.45:2181/myappB"],
        title="ZooKeeper URIs",
    )

	@validator("endpoints")
    @classmethod
    def is_comma_list_of_ips(cls, value: str) -> str | None:
        try:
            ips = [address.split(":") for address in value.split(",")]
            valid_ips = [ipaddress.ip_address(ip) for ip in ips]
        except Exception as e:
            raise ValueError(
                f"Value not a valid IP address: {e}"
            )
	
    @validator("endpoints")
    @classmethod
    def ports_are_the_same(cls, value: str) -> str | None:
        try:
            non_chroot = value.split("/")[0]
            ports = {int(address.split(":")[1]) for address in non_chroot.split(",")}
            if not len(ports) == 1:
                raise ValueError(
                    "Value does not contain the same port for every address"
                )
        except Exception as e:
            raise ValueError(
                f"Value does not contain valid ports: {e}"
            )
```

Reply:
Using `Annotated`, this is equivalent, but I might have missed your point

Reply:
Remove all the fancy validation stuff, it was causing issue with pydantic V1 compatibilty. In addition, we do not have tests any more.